flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultBatchJobRecoveryHandler.java

```diff
@@ -0,0 +1,840 @@
+
+        // get all vertices need to restart according failover strategy.
+        Set<ExecutionVertexID> verticesToReset = new HashSet<>();
+        for (ExecutionVertexID executionVertexId : missingPartitionVertices) {
+            if (!verticesToReset.contains(executionVertexId)) {
+                verticesToReset.addAll(
+                        failoverStrategy.getTasksNeedingRestart(executionVertexId, null));
+            }
+        }
+
+        resetVerticesInRecovering(verticesToReset);
+    }
+
+    private ReconcileResult reconcilePartitions(Set<ResultPartitionID> actualPartitions) {
+        Set<ResultPartitionID> expectedPartitions =
+                executionGraph.getAllIntermediateResults().values().stream()
+                        .flatMap(result -> Arrays.stream(result.getPartitions()))
+                        .filter(this::isPartitionShouldBeReconciled)
+                        .map(IntermediateResultPartition::getPartitionId)
+                        .map(this::createResultPartitionId)
+                        .collect(Collectors.toSet());
+
+        Set<ResultPartitionID> unknownPartitions =
+                Sets.difference(actualPartitions, expectedPartitions);
+        Set<ResultPartitionID> missPartitions =
+                Sets.difference(expectedPartitions, actualPartitions);
+        Set<ResultPartitionID> availablePartitions =
+                Sets.intersection(expectedPartitions, actualPartitions);
+
+        return new ReconcileResult(unknownPartitions, missPartitions, availablePartitions);
+    }
+
+    private ResultPartitionID createResultPartitionId(IntermediateResultPartitionID partitionId) {
+        final Execution producer = getProducer(partitionId).getPartitionProducer();
+        return new ResultPartitionID(partitionId, producer.getAttemptId());
+    }
+
+    private ExecutionVertex getProducer(IntermediateResultPartitionID partitionId) {
+        return executionGraph.getResultPartitionOrThrow(partitionId).getProducer();
+    }
+
+    private boolean isPartitionShouldBeReconciled(IntermediateResultPartition partition) {
+        // 1. Check if the producer of this partition is finished.
+        boolean isProducerFinished =
+                getProducer(partition.getPartitionId()).getExecutionState()
+                        == ExecutionState.FINISHED;
+
+        if (!isProducerFinished) {
+            return false;
+        }
+
+        // 2. Check if all downstream vertices for this partition are initialized.
+        boolean allConsumersInitialized =
+                partition.getIntermediateResult().getConsumerVertices().stream()
+                        .allMatch(
+                                jobVertexId -> getExecutionJobVertex(jobVertexId).isInitialized());
+
+        if (!allConsumersInitialized) {
+            return true;
+        }
+
+        // 3. If all downstream vertices are initialized, we need to check if the all partition
+        // consumers finished.
```

Comment:
Looks to me this is the minimum set of partitions to recover the progress with no loss?
But by reconciling more intermediate partitions, the job would be more robust against failures?